### DISCUSSION TITLE
ASSERT FAILED when calling tud_control_xfer during class specific requests

### DISCUSSION THREAD

**Discussion by ChrisT1111** [hathach/tinyusb#2230] [Q&A] [answered]
I have written a driver for a USB printer device. But I have an issue with the class specific requests.
When I call the function `tud_control_xfer(rhport, request, usb_dev_id, sizeof(usb_dev_id));` I get the messages: 
usbd_edpt_xfer 1243: ASSERT FAILED
tud_control_xfer 121: ASSERT FAILED

What causes these messages? Am I doing something wrong when I call tud_control_xfer?

I am testing this with the ESP32-S3 using idf version 5.0.1, tinyusb 0.15.0 and esp_tinyusb 1.3.1.
Strangely enough the calling of the function also messes up my string descriptors. 
I have posted a piece of the driver below.
```
char usb_dev_id[] ="\x00\x82MFG:TEST;MDL:Test;";
// Invoked when a control transfer occurred on an interface of this class
// Driver response accordingly to the request and the transfer stage (setup/data/ack)
// return false to stall control endpoint (e.g unsupported request)
bool printer_driver_control_xfer_cb(uint8_t rhport, uint8_t stage, tusb_control_request_t const * request)
{
  // Handle class request only
  TU_VERIFY(request->bmRequestType_bit.type == TUSB_REQ_TYPE_CLASS);

  printer_interface_t* p_printer = _printer_itf;

  if (request->bmRequestType_bit.type == TUSB_REQ_TYPE_CLASS)
  {
    //------------- Class Specific Request -------------//
    switch( request->bRequest )
    {
      case PRINTER_REQ_GET_DEVICE_ID:
        printf("GET PRINTER ID\n");
        tud_control_xfer(rhport, request, usb_dev_id, sizeof(usb_dev_id));
        break;
      default:
        printf("THIS CLASS REQUEST NOT SUPPORTED\n");
        return false; // stall unsupported request
    }
  }
  else
  {
    printf("NOT A CLASS REQUEST\n");
    return false; // stall unsupported request
  }
  return true;
}

```
The log shows the following information:

string_descr_count 5
USBD init on controller 0
sizeof(usbd_device_t) = 43
sizeof(tu_fifo_t) = 20
sizeof(tu_edpt_stream_t) = 124
dwc2->guid = 0
dwc2->gsnpsid = 4F54400A
dwc2->ghwcfg1 = 0

dwc2->ghwcfg2 = 224DD930
hw_cfg2->op_mode = 0
hw_cfg2->arch = 0
hw_cfg2->point2point = 0
hw_cfg2->hs_phy_type = 0
hw_cfg2->fs_phy_type = 0
hw_cfg2->num_dev_ep = 0
hw_cfg2->num_host_ch = 0
hw_cfg2->period_channel_support = 0
hw_cfg2->enable_dynamic_fifo = 0
hw_cfg2->mul_cpu_int = 0
hw_cfg2->nperiod_tx_q_depth = 0
hw_cfg2->host_period_tx_q_depth = 0
hw_cfg2->dev_token_q_depth = 0
hw_cfg2->otg_enable_ic_usb = 0

dwc2->ghwcfg3 = C804B5
hw_cfg3->xfer_size_width = 0
hw_cfg3->packet_size_width = 0
hw_cfg3->otg_enable = 0
hw_cfg3->i2c_enable = 0
hw_cfg3->vendor_ctrl_itf = 0
hw_cfg3->optional_feature_removed = 0
hw_cfg3->synch_reset = 0
hw_cfg3->otg_adp_support = 0
hw_cfg3->otg_enable_hsic = 0
hw_cfg3->battery_charger_support = 0
hw_cfg3->lpm_mode = 0
hw_cfg3->total_fifo_size = 0

dwc2->ghwcfg4 = D3F0A030
hw_cfg4->num_dev_period_in_ep = 0
hw_cfg4->power_optimized = 0
hw_cfg4->ahb_freq_min = 0
hw_cfg4->hibernation = 0
hw_cfg4->service_interval_mode = 0
hw_cfg4->ipg_isoc_en = 0
hw_cfg4->acg_enable = 0
hw_cfg4->utmi_phy_data_width = 0
hw_cfg4->dev_ctrl_ep_num = 0
hw_cfg4->iddg_filter_enabled = 0
hw_cfg4->vbus_valid_filter_enabled = 0
hw_cfg4->a_valid_filter_enabled = 0
hw_cfg4->b_valid_filter_enabled = 0
hw_cfg4->dedicated_fifos = 0
hw_cfg4->num_dev_in_eps = 0
hw_cfg4->dma_desc_enable = 0
hw_cfg4->dma_dynamic = 0
Fullspeed PHY init
USBD Bus Reset : Full Speed

USBD Setup Received 80 06 00 01 00 00 40 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes
USBD Bus Reset : Full Speed
IN RESET

USBD Setup Received 00 05 0B 00 00 00 00 00 
  Set Address
USBD Xfer Complete on EP 80 with 0 bytes

USBD Setup Received 80 06 00 01 00 00 12 00
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 02 00 00 FF 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 32 bytes ...
USBD Xfer Complete on EP 80 with 32 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 03 03 09 04 FF 00 
  Get Descriptor String[3]
  Queue EP 80 with 30 bytes ...
USBD Xfer Complete on EP 80 with 30 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 03 00 00 FF 00
  Get Descriptor String[0]
  Queue EP 80 with 4 bytes ...
USBD Xfer Complete on EP 80 with 4 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 02 03 09 04 FF 00
  Get Descriptor String[2]
  Queue EP 80 with 46 bytes ...
USBD Xfer Complete on EP 80 with 46 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 06 00 00 0A 00
  Get Descriptor Device Qualifier
  Stall EP0

USBD Setup Received 80 06 00 01 00 00 12 00
  Get Descriptor Device
  Queue EP 80 with 18 bytes ...
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 02 00 00 09 01
  Get Descriptor Configuration[0]
  Queue EP 80 with 32 bytes ...
USBD Xfer Complete on EP 80 with 32 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 00 09 01 00 00 00 00 00
  Set Configuration
  Open EP 02 with Size = 64
  Open EP 81 with Size = 64
    Allocated 64 bytes at offset 896  Queue EP 02 with 64 bytes ...
  Bind EP 02 to driver id 0
  Bind EP 81 to driver id 0
  Queue EP 80 with 0 bytes ...
USBD Xfer Complete on EP 80 with 0 bytes

USBD Setup Received 80 06 03 03 09 04 03 01
  Get Descriptor String[3]
  Queue EP 80 with 30 bytes ...
USBD Xfer Complete on EP 80 with 30 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received A1 00 00 00 00 00 F1 03
  PRINTER control request
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
USBD Xfer Complete on EP 80 with 21 bytes
  PRINTER control complete
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
usbd_edpt_xfer 1254: ASSERT FAILED
tud_control_xfer 126: ASSERT FAILED

USBD Setup Received A1 00 00 00 00 00 00 04 
  PRINTER control request
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
USBD Xfer Complete on EP 80 with 21 bytes
  PRINTER control complete
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
usbd_edpt_xfer 1254: ASSERT FAILED
tud_control_xfer 126: ASSERT FAILED

USBD Setup Received A1 00 00 00 00 00 00 04 
  PRINTER control request
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
USBD Xfer Complete on EP 80 with 21 bytes
  PRINTER control complete
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
usbd_edpt_xfer 1254: ASSERT FAILED
tud_control_xfer 126: ASSERT FAILED

USBD Setup Received A1 00 00 00 00 00 00 04
  PRINTER control request
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
USBD Xfer Complete on EP 80 with 21 bytes
  PRINTER control complete
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
usbd_edpt_xfer 1254: ASSERT FAILED
tud_control_xfer 126: ASSERT FAILED

USBD Setup Received A1 00 00 00 00 00 00 04
  PRINTER control request
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
USBD Xfer Complete on EP 80 with 21 bytes
  PRINTER control complete
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes
GET PRINTER ID
  Queue EP 80 with 21 bytes ...
usbd_edpt_xfer 1254: ASSERT FAILED
tud_control_xfer 126: ASSERT FAILED

USBD Setup Received 80 06 00 03 00 00 FF 00 
  Get Descriptor String[0]
  Queue EP 80 with 4 bytes ...
USBD Xfer Complete on EP 80 with 4 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 01 03 09 04 FF 00 
  Get Descriptor String[1]
  Queue EP 80 with 10 bytes ...
USBD Xfer Complete on EP 80 with 10 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 02 03 09 04 FF 00
  Get Descriptor String[2]
  Queue EP 80 with 46 bytes ...
USBD Xfer Complete on EP 80 with 46 bytes
  Queue EP 00 with 0 bytes ...
USBD Xfer Complete on EP 00 with 0 bytes

Does anyone have an idea what I am doing wrong?

**Comment by kkitayam** [comment]
> USBD Setup Received A1 00 00 00 00 00 F1 03
> PRINTER control request
> GET PRINTER ID
> Queue EP 80 with 21 bytes ...
> USBD Xfer Complete on EP 80 with 21 bytes
> PRINTER control complete
> GET PRINTER ID
> Queue EP 80 with 21 bytes ...
> Queue EP 00 with 0 bytes ...

 It seems that ZLP of 1st GET PRINTER ID is missing.
Does the driver send ZLP as soon as possible? If host timed out of waiting for ZLP, The host may resend the same request. I guess it makes handshake messed up.

**Reply by ChrisT1111** [reply]
Thank you for your reply.
Currently I do not specifically send a zero length packet anywhere. Is this not something the library does?
The only thing I do is call tud_control_xfer(rhport, request, usb_dev_id, sizeof(usb_dev_id)); when I receive a GET_PRINTER_ID request.
I based the driver on the cdc_device.c class driver and the printer USB class definition document (https://www.usb.org/sites/default/files/usbprint11a021811.pdf).
I don't see a ZLP packet in the cdc_device code.
I am not that experienced in the USB standard. When should I send a zero length packet. Is there any documentation I could read on this subject in relation to the USB class definition?

**Reply by kkitayam** [reply]
> Currently I do not specifically send a zero length packet anywhere. Is this not something the library does?

Ah, Yes. When `control_xfer_cb` returns true, the device stack sends a ZLP. Sorry for confusion.

But, your log shows ZLP is missing. I suspect `control_xfer_cb` was delayed by something then the host timed out waiting for a ZLP.
If your implementation uses RTOS, the task priority of the TinyUSB stack task should be higher than other tasks to avoid `control_xfer_cb` delays.
Simply suppressing log output may improve this. Log output is very slow with UART.

**Reply by ChrisT1111** [reply]
I do use FreeRTOS. To check if the priority is the issue, I made a new project with only the USB functionality and set the task priority high. I also remove my print statements and decreased the TinyUSB log level to zero. Unfortunately the issue remains. Is it possible that there is a bug in the library that causes the issue? Any ideas on what I could check?

**Comment by Rocky04** [comment, marked as answer]
tl;:dr

The message "tud_control_xfer 121: ASSERT FAILED" likely means that the data state transfer failed and so [the assert for it](https://github.com/hathach/tinyusb/blob/46f7cf4da2959b96d3af6b7f23d0d02ff7187f2f/src/device/usbd_control.c#L123) too. Your code is called for each stage of the setup request, so at least one time for `CONTROL_STAGE_SETUP` and one time for `CONTROL_STAGE_ACK` and maybe even for `CONTROL_STAGE_DATA`. So maybe the assert fails because you want to return data after the acknowledged status and it fails because this is the status stage and so the final one which completes the transfer.

I also disagree that the TinyUSB stack takes care of the ZLP package. For example on the HID class driver these are send on the setup stage. [Like here](https://github.com/hathach/tinyusb/blob/46f7cf4da2959b96d3af6b7f23d0d02ff7187f2f/src/class/hid/hid_device.c#L360)... They are also present on the CDC example, [like here](https://github.com/hathach/tinyusb/blob/46f7cf4da2959b96d3af6b7f23d0d02ff7187f2f/src/class/cdc/cdc_device.c#L409). _When these are missing it don't work anymore._ Again to be clear, if you are sending data you don't need to send a ZLP. The TinyUSB stack sends a ZLP in the case you are sending as much data as the endpoint can handle. For example a HID transfer can be bigger than the endpoint can handle in a single transaction. Then the transfer will be split into multiple transactions. Due to short packages the host don't know if the device wants to send more data or if it already sent all. So the host will schedule another transaction and if the device didn't have any more data to send it will send a ZLP. Your application only sends the transfer regardless how much data the endpoint can handle while TinyUSB takes care of it.

If the control transfer function returns false TinyUSB would just stall the request, so likely returning true only prevents the STALL.

_Edit:_
About the issue, I use this code for requests where I only care about the setup stage. It will prevent a stall for the supported request because it will return true in the case the request is supported but for a stage where I'm not interested in. Then I either send data for the data state or a ZLP via `tud_control_status(rhport, request)` for an ingoing transfer, or prepare to receive data for an outgoing transfer. _Like there must always be a data stage, either by receiving data or sending and that must be at least a ZLP. Unless of course the host cancels the setup request, by sending another one._
```
case PRINTER_REQ_GET_DEVICE_ID:
  TU_VERIFY(CONTROL_STAGE_SETUP == stage, true);
  ...
```

_Edit 2:_
That the assert failed the third time also indicate that this is a stage issue. The first time should be for the setup stage when that was received, the second is likely for the data stage when the data was sent and the third is maybe for the acknowledgement of the status stage when the host should completed the request.

_Edit 3:_
The second assert failed because of the first one on a deeper level of the stack trace. Likely because you want to send data [while the control endpoint is still busy](https://github.com/hathach/tinyusb/blob/46f7cf4da2959b96d3af6b7f23d0d02ff7187f2f/src/device/usbd.c#L1249) due to a previous send attempt. The busy should only be released when another setup request occurred or when the previous transfer for the endpoint was completed.

According to the log it seems to me you are trying to send three times 21 bytes to the host for the same setup request. It was successfully sent the first time but then you try to send it a second time while the host didn't received it and so it failed the third time as the endpoint is still busy because the second didn't finished.

_Edit 4:_
I can recommend the book "USB Complete". There it's mentioned that the data stage is optional and some requests have it while others not. But the Get_Descriptor has a data stage where the device needs to send data even that is just a ZLP. The Set_Configuration on the other hand don't has a data stage.

**Reply by ChrisT1111** [reply]
Thank you. This was indeed the problem. I did send the device id on every stage of the request.
I edited the code to send the ID only in the setup stage and know the code works. In all other cases I return true. Thanks for the book tip. I will certainly take a look at it.
